src/lazy/binary/immutable_buffer.rs

```diff
@@ -157,6 +159,257 @@ impl<'a> ImmutableBuffer<'a> {
+            let remaining = self.consume(flex_uint.size_in_bytes());
+            return Ok((flex_uint, remaining));
+        }
+
+        // If we reach this point, the first byte was a zero. The FlexUInt is at least 9 bytes in size.
+        // We need to inspect the second byte to see how many more prefix bits there are.
+        if bytes_available < 2 {
+            return incomplete();
+        }
+        let second_byte = bytes[1];
+
+        if second_byte & 0b11 == 0b00 {
+            // The flag bits in the second byte indicate at least two more bytes, meaning the total
+            // length is more than 10 bytes. We're not equipped to handle this.
+            return IonResult::decoding_error("found a >10 byte VarUInt too large to fit in a u64");
+        }
+
+        if second_byte & 0b11 == 0b10 {
+            // The lowest bit of the second byte is empty, the next lowest is not. The encoding
+            // is 10 bytes.
+
+            if bytes_available < 10 {
+                return incomplete();
+            }
+
+            let flex_uint = self.read_10_byte_flex_primitive_as_uint(
+                support_sign_extension,
+                bytes,
+                second_byte,
+            )?;
+            let remaining = self.consume(flex_uint.size_in_bytes());
+            return Ok((flex_uint, remaining));
+        }
+
+        // The lowest bit of the second byte is set. The encoding is 9 bytes.
+        if bytes_available < 9 {
+            return incomplete();
+        }
+        // There are 57-63 bits of magnitude. We can decode the remaining bytes in a u64.
+        let remaining_data = &bytes[1..9];
+        // We know that the slice is 8 bytes long, so we can unwrap() the conversion to [u8; 8]
+        // Lop off the lowest bit to discard the `end` flag.
+        let value = u64::from_le_bytes(remaining_data[..8].try_into().unwrap()) >> 1;
+        let flex_uint = FlexUInt::new(9, value);
+        let remaining_input = self.consume(flex_uint.size_in_bytes());
+        Ok((flex_uint, remaining_input))
+    }
+
+    /// Helper method to handle flex primitives whose encoding requires 10 bytes. This case is
+    /// complex because it requires evaluating data beyond the supported 64 bits of representation
+    /// to detect overflow and support signed re-interpretation.
+    fn read_10_byte_flex_primitive_as_uint(
+        self,
+        support_sign_extension: bool,
+        bytes: &'a [u8],
+        second_byte: u8,
+    ) -> IonResult<FlexUInt> {
+        // There are 10 prefix (continuation) bits, 64 bits of magnitude, and 6 bits of sign
+        // extension (if enabled). We cannot store the highest 6 bits, so this method just checks
+        // to make sure that they do not modify the meaning of the value in the lower 64 bits.
+        // For signed values, this means the 6 extra bits must all be the same as the 64th bit.
+        // For unsigned values, this means that the 6 extra bits must all be `1`.
```

Comment:
Hold on... why are the extra bits always `1` for unsigned values? Why not `0`?

Reply:
Oops! The code does the right thing, but this comment got it backwards. Will fix, thanks.